components/eamxx/src/physics/mam/eamxx_mam_aci_process_interface.cpp

```diff
@@ -0,0 +1,676 @@
+
+  //---------------------------------------------------------------------------------
+  // Diagnotics variables from the hetrozenous ice nucleation scheme
+  //---------------------------------------------------------------------------------
+
+  for(int i = 0; i < hetro_scratch_; ++i)
+    Kokkos::resize(diagnostic_scratch_[i], ncol_, nlev_);
+
+  //---------------------------------------------------------------------------------
+  // Initialize the processes
+  //---------------------------------------------------------------------------------
+
+  mam4::AeroConfig aero_config;
+  // configure the nucleation parameterization
+  mam4::NucleateIce::Config nucleate_ice_config;
+  nucleate_ice_.init(aero_config, nucleate_ice_config);
+
+  // configure the heterogeneous freezing parameterization
+  mam4::Hetfrz::Config hetfrz_config;
+  hetfrz_.init(aero_config, hetfrz_config);
+
+  //---------------------------------------------------------------------------------
+  // Setup preprocessing and post processing
+  //---------------------------------------------------------------------------------
+  // set up our preprocess  and postprocess functors
+  preprocess_.initialize(ncol_, nlev_, wet_atm_, wet_aero_, dry_atm_,
+                         dry_aero_);
+
+  postprocess_.initialize(ncol_, nlev_, wet_atm_, wet_aero_, dry_atm_,
+                          dry_aero_);
+
+}  // end function initialize_impl
+
+// ================================================================
+//  RUN_IMPL
+// ================================================================
+void MAMAci::run_impl(const double dt) {
+  const auto scan_policy = ekat::ExeSpaceUtils<
+      KT::ExeSpace>::get_thread_range_parallel_scan_team_policy(ncol_, nlev_);
+
+  // preprocess input -- needs a scan for the calculation of local derivied
+  // quantities
+  Kokkos::parallel_for("preprocess", scan_policy, preprocess_);
+  Kokkos::fence();
+
+  haero::ThreadTeamPolicy team_policy(ncol_, Kokkos::AUTO);
+
+  // FIXME: Temporary assignment of nc
+  mam_coupling::copy_view_lev_slice(team_policy, wet_atm_.nc, nlev_,  // inputs
+                                    nc_inp_to_aci_);                  // output
+
+  compute_w0_and_rho(team_policy, dry_atm_, top_lev_, nlev_,
+                     // output
+                     w0_, rho_);
+
+  // Get w_sec_int_ from w_sec_mid_
+  compute_values_at_interfaces(team_policy, w_sec_mid_, dry_atm_.dz, nlev_,
+                               // output
+                               w_sec_int_);
+
+  compute_tke_using_w_sec(team_policy, w_sec_int_, nlev_,
+                          // output
+                          tke_);
+
+  Kokkos::fence();  // wait for for tke_ to be computed.
+  compute_subgrid_scale_velocities(team_policy, tke_, wsubmin_, top_lev_, nlev_,
+                                   // output
+                                   wsub_, wsubice_, wsig_);
+
+  Kokkos::fence();  // wait for wsig_ to be computed.
```

Comment:
Does aitken diameter need the result from the previous kernel? If not, you don't need the fence. Similar for other fences: only add them if you _need_ the previous kernel(s) to complete before the next starts.